File: src/main/services/addy/pairing.ts
import { AddyError, type AddySidecar } from './sidecar'

/**
 * Driving a pairing from main.
 *
 * Three parties: this device's sidecar, which holds the keys and runs SPAKE2;
 * the relay, which carries opaque frames between two mailboxes and learns
 * nothing; and the user, who compares seven emoji with somebody in the room or
 * on the phone.
 *
 * THIS MODULE HOLDS NO KEY AND MAKES NO DECISION. It moves frames and reports
 * what the sidecar said. The two things that matter -- that the emoji appear
 * only after a key confirmation verifies, and that three wrong codes end the
 * session -- are both enforced in the sidecar, where a bug here cannot reach
 * them.
 */

/** The unauthenticated rendezvous is a poll, so this is how long to keep
 *  asking before giving up on the other device. Long: the person on the far
 *  end is typing a six-word code, and a pairing that timed out while they were
 *  still reading it is one they blame on the product. */
const RENDEZVOUS_TIMEOUT_MS = 180_000

export interface PairingHandle {
  /** Shown on this screen and read to the other person. Never sent anywhere. */
  code: string
  pairingId: string
}

export interface PairingConfirmation {
  /** Seven emoji, identical on both devices. */
  sas: string[]
  /** The same list as words, so a phone call works as well as a photograph. */
  sasWords: string
  peer: { pubSign: string; pubEnc?: string }
  /** The joining device's own public halves, present only on the joiner, for
   *  the initiator to write into the roster entry that adds it. */
  self?: { pubSign: string; pubEnc: string }
}

export interface PairingDeps {
  addyd: AddySidecar
  /** `https://relay.example`. The rendezvous is UNAUTHENTICATED -- a joining
   *  device has no account yet, which is the whole point -- so this takes a
   *  bare URL rather than a signed client. */
  baseURL: string
}

/** Posts a frame into the peer's mailbox. */
async function publish(deps: PairingDeps, id: string, role: string, frame: unknown): Promise<void> {
  const resp = await fetch(`${deps.baseURL}/v1/pair/${id}?as=${role}`, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ frame: Buffer.from(JSON.stringify(frame), 'utf8').toString('base64') })
  })
  if (!resp.ok) {
    throw new AddyError('relay-unreachable', `the relay would not carry the frame (${resp.status})`)
  }
}

/**
 * Waits for a frame in this role's mailbox.
 *
 * The relay answers 204 when nothing arrived within its own poll window, so
 * this loops. An implementation that treated 204 as an error would give up the
 * first time the other person took more than twenty seconds to type.
 */
async function receive<T>(deps: PairingDeps, id: string, role: string, signal: AbortSignal): Promise<T> {
  const deadline = Date.now() + RENDEZVOUS_TIMEOUT_MS
  while (Date.now() < deadline) {
    if (signal.aborted) throw new AddyError('pairing-expired', 'the pairing was cancelled')
    const resp = await fetch(`${deps.baseURL}/v1/pair/${id}?as=${role}`, { signal }).catch(() => null)
    if (!resp) throw new AddyError('relay-unreachable', 'the relay stopped answering')
    if (resp.status === 204) continue
    if (!resp.ok) {
      throw new AddyError('relay-unreachable', `the relay refused the rendezvous (${resp.status})`)
    }
    const { frame } = (await resp.json()) as { frame: string }
    return JSON.parse(Buffer.from(frame, 'base64').toString('utf8')) as T
  }
  throw new AddyError('pairing-expired', 'the other device did not answer in time')
}

/**
 * The device SHOWING the code.
 *
 * Returns as soon as there is a code to display, and the rest happens in the
 * promise it also returns. That split matters: the user has to be able to read
 * the code out loud while this is still waiting for the other end, and an API
 * that only resolved once pairing completed would have nothing to show them.
 */
export function beginPairing(
  deps: PairingDeps,
  signal: AbortSignal
): Promise<{ handle: PairingHandle; confirmed: Promise<PairingConfirmation> }> {
  return deps.addyd
    .send<{ code: string; pairingId: string; startFrame: unknown }>('pairBegin')
    .then((begun) => {
      const handle: PairingHandle = { code: begun.code, pairingId: begun.pairingId }

      const confirmed = (async (): Promise<PairingConfirmation> => {
        await publish(deps, begun.pairingId, 'initiator', begun.startFrame)

        const reply = await receive<{
          msgB: string
          confirmB: string
          pubSign: string
          pubEnc: string
        }>(deps, begun.pairingId, 'initiator', signal)

        // The sidecar verifies the confirmation MAC and only then derives the
        // emoji. If it throws here, the other end did not prove it knows the
        // code -- and the emoji are never computed at all.
        const replied = await deps.addyd.send<PairingConfirmation & { confirmFrame: unknown }>(
          'pairReply',
          { pairingId: begun.pairingId, ...reply }
        )
        await publish(deps, begun.pairingId, 'initiator', replied.confirmFrame)

        return { sas: replied.sas, sasWords: replied.sasWords, peer: replied.peer }
      })()

      return { handle, confirmed }
    })
}

/**
 * The device TYPING the code.
 *
 * It needs the pairing id as well, and that is not an oversight in the
 * protocol -- it is a property of it. The id is inside the start frame, and
 * the frame sits in a mailbox addressed BY the id, so a joiner cannot fetch it
 * without already knowing it. The two therefore travel together: the initiator
 * shows both, by QR or as one string, and the panel splits them.
 *
 * The consequence is worth naming, because it is a feature: a code on its own
 * is useless later. Two devices have to be pairing AT THE SAME TIME, so a code
 * overheard this morning cannot be redeemed this afternoon.
 */
export async function joinPairing(
  deps: PairingDeps,
  code: string,
  pairingId: string,
  signal: AbortSignal
): Promise<PairingConfirmation> {
  const start = await receive<{ pairingId: string; msgA: string; pubSign: string }>(
    deps,
    pairingId,
    'joiner',
    signal
  )
  const joined = await deps.addyd.send<{ replyFrame: unknown }>('pairJoin', {
    code,
    pairingId,
    msgA: start.msgA,
    pubSign: start.pubSign
  })
  await publish(deps, pairingId, 'joiner', joined.replyFrame)

  const confirm = await receive<{ confirmA: string }>(deps, pairingId, 'joiner', signal)
  return deps.addyd.send<PairingConfirmation>('pairConfirm', { pairingId, confirmA: confirm.confirmA })
}

/** Ends a session, forgetting the shared secret. Called when the user says the
 *  emoji do not match, and on every ordinary completion. */
export async function forgetPairing(deps: PairingDeps, pairingId: string): Promise<void> {
  await deps.addyd.send('pairForget', { pairingId }).catch(() => undefined)
}

File: sidecar/addyd/pairing.go
package main

import (
	"crypto/ed25519"
	"encoding/base64"
	"encoding/hex"
	"sync"

	"github.com/opsmaxx/opsmaxx/sidecar/addyd/pair"
	"github.com/opsmaxx/opsmaxx/sidecar/addyd/protocol"
)

// Pairing, in the process that holds the keys.
//
// The relay carries opaque frames between the two devices and learns nothing.
// What it CANNOT do, and what makes the short code safe, is guess: SPAKE2 turns
// a six-word code into a shared secret in one round trip, and a wrong guess
// gives an attacker one attempt rather than an offline target.
//
// THE ATTEMPT LIMITER IS ON THE DEVICE, not the server. A server-side limiter
// protects a server that is not the thing under attack -- the code is between
// two devices, and a relay that counted attempts would be a relay trusted to
// count honestly.

// pairings holds in-flight sessions. In memory only, like everything else
// here: a pairing that survived a restart would be a pairing whose other end
// has already given up.
var pairings = struct {
	mu        sync.Mutex
	initiator map[string]*pair.Initiator
	joiner    map[string]*pair.Joiner
	// attempts counts failed joins per pairing id. THE LIMITER, and it lives
	// beside the sessions rather than in the parent because a limiter the
	// parent enforced would be one a bug in the parent could skip.
	attempts map[string]int
}{
	initiator: map[string]*pair.Initiator{},
	joiner:    map[string]*pair.Joiner{},
	attempts:  map[string]int{},
}

// maxAttempts is how many wrong codes a pairing tolerates before it is dead.
//
// Three. SPAKE2 gives an attacker one guess per attempt and no offline target,
// so the limit is not protecting the code's entropy -- it is making a
// shoulder-surfed or mistyped code fail visibly rather than letting somebody
// grind quietly while the user wonders why pairing is slow.
const maxAttempts = 3

// --- the device showing the code ---

// handlePairBegin mints a code and returns the first frame.
func handlePairBegin(req Request) (any, error) {
	keys.mu.RLock()
	device := keys.device
	loaded := keys.loaded
	keys.mu.RUnlock()
	if !loaded || device == nil {
		return nil, codedf(ErrNotPaired, "no account is loaded")
	}

	code, err := pair.NewCode()
	if err != nil {
		return nil, wrapCoded(ErrInternal, err, "minting a pairing code")
	}
	pubSign := device.Sign.Public().(ed25519.PublicKey)
	initiator, frame, err := pair.Begin(code, pubSign)
	if err != nil {
		return nil, wrapCoded(ErrInternal, err, "starting a pairing")
	}

	id := hex.EncodeToString(frame.PairingID)
	pairings.mu.Lock()
	pairings.initiator[id] = initiator
	pairings.mu.Unlock()

	return map[string]any{
		// The code is shown to the user and read aloud or typed on the other
		// device. It never goes over the relay.
		"code":      code,
		"pairingId": id,
		"startFrame": map[string]string{
			"pairingId": id,
			"msgA":      base64.StdEncoding.EncodeToString(frame.MsgA),
			"pubSign":   hex.EncodeToString(frame.PubSign),
		},
	}, nil
}

// handlePairReply takes the joiner's frame and returns the confirmation plus
// the SAS.
//
// THE SAS IS RETURNED ONLY AFTER THE KEY CONFIRMATION MAC VERIFIES. That
// ordering is the whole value of showing emoji: a SAS computed before
// confirmation would be a SAS an attacker can influence, and a user comparing
// two matching lists would be confirming the attacker's session.
func handlePairReply(req Request) (any, error) {
	var in struct {
		PairingID string `json:"pairingId"`
		MsgB      string `json:"msgB"`
		ConfirmB  string `json:"confirmB"`
		PubSign   string `json:"pubSign"`
		PubEnc    string `json:"pubEnc"`
	}
	if err := decodeParams(req, &in); err != nil {
		return nil, err
	}

	pairings.mu.Lock()
	initiator, ok := pairings.initiator[in.PairingID]
	attempts := pairings.attempts[in.PairingID]
	pairings.mu.Unlock()
	if !ok {
		return nil, codedf(ErrPairingExpired, "no pairing is in progress with that id")
	}
	if attempts >= maxAttempts {
		return nil, codedf(ErrPairingRefused,
			"too many failed attempts on this pairing; start a new one")
	}

	msgB, err1 := base64.StdEncoding.DecodeString(in.MsgB)
	confirmB, err2 := base64.StdEncoding.DecodeString(in.ConfirmB)
	pubSign, err3 := hex.DecodeString(in.PubSign)
	pubEnc, err4 := hex.DecodeString(in.PubEnc)
	if err1 != nil || err2 != nil || err3 != nil || err4 != nil {
		return nil, codedf(ErrConfigInvalid, "the reply frame is malformed")
	}

	confirm, err := initiator.Reply(pair.ReplyFrame{
		MsgB: msgB, ConfirmB: confirmB, PubSign: pubSign, PubEnc: pubEnc,
	})
	if err != nil {
		// A wrong code lands here, as a failed MAC. Counted, and counted
		// before the error is returned so a caller that retries in a loop
		// still runs out.
		pairings.mu.Lock()
		pairings.attempts[in.PairingID]++
		spent := pairings.attempts[in.PairingID] >= maxAttempts
		if spent {
			delete(pairings.initiator, in.PairingID)
		}
		pairings.mu.Unlock()
		return nil, wrapCoded(ErrSASMismatch, err, "the other device did not confirm")
	}

	sas, err := initiator.SAS()
	if err != nil {
		return nil, wrapCoded(ErrInternal, err, "deriving the comparison emoji")
	}

	return map[string]any{
		"confirmFrame": map[string]string{
			"confirmA": base64.StdEncoding.EncodeToString(confirm.ConfirmA),
		},
		"sas":      sas.Emoji[:],
		"sasWords": sas.Words(),
		"peer": map[string]string{
			"pubSign": hex.EncodeToString(initiator.PeerPubSign()),
			"pubEnc":  hex.EncodeToString(pubEnc),
		},
	}, nil
}

// --- the device typing the code ---

// handlePairJoin answers the first frame.
func handlePairJoin(req Request) (any, error) {
	var in struct {
		Code      string `json:"code"`
		PairingID string `json:"pairingId"`
		MsgA      string `json:"msgA"`
		PubSign   string `json:"pubSign"`
	}
	if err := decodeParams(req, &in); err != nil {
		return nil, err
	}

	keys.mu.RLock()
	device := keys.device
	keys.mu.RUnlock()
	if device == nil {
		// A joining device has its own key before it has an account: it is
		// about to be ADDED to one. So the key is minted here rather than
		// required to exist.
		fresh, err := protocol.NewDeviceKeys()
		if err != nil {
			return nil, wrapCoded(ErrInternal, err, "minting a device key")
		}
		keys.mu.Lock()
		keys.device = fresh
		keys.mu.Unlock()
		device = fresh
	}

	pairingID, err1 := hex.DecodeString(in.PairingID)
	msgA, err2 := base64.StdEncoding.DecodeString(in.MsgA)
	peerSign, err3 := hex.DecodeString(in.PubSign)
	if err1 != nil || err2 != nil || err3 != nil {
		return nil, codedf(ErrConfigInvalid, "the start frame is malformed")
	}

	joiner, reply, err := pair.Join(in.Code, pair.StartFrame{
		PairingID: pairingID, MsgA: msgA, PubSign: peerSign,
	}, device.Sign.Public().(ed25519.PublicKey), device.Enc.PublicKey().Bytes())
	if err != nil {
		return nil, wrapCoded(ErrPairingRefused, err, "joining the pairing")
	}

	pairings.mu.Lock()
	pairings.joiner[in.PairingID] = joiner
	pairings.mu.Unlock()

	return map[string]any{
		"replyFrame": map[string]string{
			"msgB":     base64.StdEncoding.EncodeToString(reply.MsgB),
			"confirmB": base64.StdEncoding.EncodeToString(reply.ConfirmB),
			"pubSign":  hex.EncodeToString(reply.PubSign),
			"pubEnc":   hex.EncodeToString(reply.PubEnc),
		},
		// Deliberately NO SAS here. The joiner has not verified the
		// initiator's confirmation yet, so any emoji it showed now would be
		// emoji derived from a session nobody has authenticated.
	}, nil
}

// handlePairConfirm verifies the initiator's MAC and only then returns the SAS.
func handlePairConfirm(req Request) (any, error) {
	var in struct {
		PairingID string `json:"pairingId"`
		ConfirmA  string `json:"confirmA"`
	}
	if err := decodeParams(req, &in); err != nil {
		return nil, err
	}

	pairings.mu.Lock()
	joiner, ok := pairings.joiner[in.PairingID]
	pairings.mu.Unlock()
	if !ok {
		return nil, codedf(ErrPairingExpired, "no pairing is in progress with that id")
	}

	confirmA, err := base64.StdEncoding.DecodeString(in.ConfirmA)
	if err != nil {
		return nil, codedf(ErrConfigInvalid, "the confirm frame is malformed")
	}
	if err := joiner.Confirm(pair.ConfirmFrame{ConfirmA: confirmA}); err != nil {
		pairings.mu.Lock()
		pairings.attempts[in.PairingID]++
		if pairings.attempts[in.PairingID] >= maxAttempts {
			delete(pairings.joiner, in.PairingID)
		}
		pairings.mu.Unlock()
		return nil, wrapCoded(ErrSASMismatch, err, "the other device did not confirm")
	}

	sas, err := joiner.SAS()
	if err != nil {
		return nil, wrapCoded(ErrInternal, err, "deriving the comparison emoji")
	}

	keys.mu.RLock()
	device := keys.device
	keys.mu.RUnlock()

	return map[string]any{
		"sas":      sas.Emoji[:],
		"sasWords": sas.Words(),
		"peer": map[string]string{
			"pubSign": hex.EncodeToString(joiner.PeerPubSign()),
		},
		// The joiner's own public halves, which the initiator needs in order
		// to write the roster entry that adds it.
		"self": map[string]string{
			"pubSign": hex.EncodeToString(device.Sign.Public().(ed25519.PublicKey)),
			"pubEnc":  hex.EncodeToString(device.Enc.PublicKey().Bytes()),
		},
	}, nil
}

// handlePairForget drops a session.
//
// Called when the user says the emoji do not match, and on every ordinary
// completion. A pairing left in the map is a session an attacker can still
// send frames to, and the map is the only thing holding the shared secret.
func handlePairForget(req Request) (any, error) {
	var in struct {
		PairingID string `json:"pairingId"`
	}
	if err := decodeParams(req, &in); err != nil {
		return nil, err
	}
	pairings.mu.Lock()
	delete(pairings.initiator, in.PairingID)
	delete(pairings.joiner, in.PairingID)
	delete(pairings.attempts, in.PairingID)
	pairings.mu.Unlock()
	return map[string]any{"ok": true}, nil
}
